File: app/src/main/java/com/tablehockey/game/network/Links.kt
package com.tablehockey.game.network

import org.json.JSONObject

/**
 * Transport-agnostic ends of a two-device match. WiFi (TCP + NSD) and
 * Bluetooth (RFCOMM) both speak the same line-delimited JSON protocol from
 * [NetCodec], so the game screen only ever sees these two interfaces.
 */
interface HostLink {
    interface Listener {
        fun onClientConnected()
        fun onClientDisconnected()
        fun onMessage(obj: JSONObject)
    }

    var listener: Listener?

    /** Snapshots per second this transport comfortably carries. */
    val stateHz: Int

    fun send(json: String)
    fun stop()
}

interface GuestLink {
    interface Listener {
        fun onConnected()
        fun onConnectFailed(reason: String)
        fun onDisconnected()
        fun onMessage(obj: JSONObject)
    }

    var listener: Listener?

    /** Input packets per second this transport comfortably carries. */
    val inputHz: Int

    fun send(json: String)
    fun disconnect()
}

File: ios/PowerPlayHockeyCore/Sources/PowerPlayHockeyCore/Network/Links.swift
import Foundation

/// Transport-agnostic delegate for receiving host network events.
public protocol HostLinkListener: AnyObject {
    func onClientConnected()
    func onClientDisconnected()
    func onMessage(_ json: [String: Any])
}

/// Transport-agnostic host side of a multiplayer hockey match.
public protocol HostLink: AnyObject {
    var listener: HostLinkListener? { get set }
    var stateHz: Int { get }
    func send(_ json: String)
    func stop()
}

/// Transport-agnostic delegate for receiving guest network events.
public protocol GuestLinkListener: AnyObject {
    func onConnected()
    func onConnectFailed(reason: String)
    func onDisconnected()
    func onMessage(_ json: [String: Any])
}

/// Transport-agnostic guest side of a multiplayer hockey match.
public protocol GuestLink: AnyObject {
    var listener: GuestLinkListener? { get set }
    var inputHz: Int { get }
    func send(_ json: String)
    func disconnect()
}
